src/engine/Join.cpp

```diff
@@ -843,3 +823,29 @@ void Join::hashJoin(const IdTable& dynA, size_t jc1, const IdTable& dynB,
              << ", size = " << dynRes->size() << "\n";
 }
 
+// ___________________________________________________________________________ 
+template <int A_WIDTH, int B_WIDTH, int TABLE_WIDTH>
+void Join::addCombinedRowToIdTable(
+  const std::decay_t<typename IdTableStatic<A_WIDTH>::const_row_type>& rowA,
+  const std::decay_t<typename IdTableStatic<B_WIDTH>::const_row_type>& rowB,
+  const size_t jcRowB,
+  IdTableStatic<TABLE_WIDTH>* table) const {
+  // Add a new, empty row.
+  const size_t backIndex = table->size();
+  table->push_back();
+  
+  // Copy the entire rowA in the table.
+  for (size_t h = 0; h < rowA.size(); h++) {
+    (*table)(backIndex, h) = rowA[h];
+  }
+
+  // Copy rowB columns before the join column.
+  for (size_t h = 0; h < jcRowB; h++) {
+    (*table)(backIndex, h + rowA.size()) = rowB[h];
+  }
+
+  // Copy rowB columns after the join column.
+  for (size_t h = jcRowB + 1; h < rowB.size(); h++) {
+    (*table)(backIndex, h + rowA.size() - 1) = rowB[h];
+  }
+}
```

Comment:
Can you check, if you can also call and use this function from inside the "normal" zipper join?
Then we can also get rid of this awful code duplication there.

Reply:
I tried, but I get a rather weird error:
```
qlever-code/src/engine/Join.cpp: In instantiation of ‘static void Join::doGallopInnerJoin(TagType, IdTableView<WIDTH>&, size_t, IdTableView<R_WIDTH>&, size_t, IdTableStatic<OUT_WIDTH>*) [with TagType = Join::LeftLargerTag; int L_WIDTH = 0; int R_WIDTH = 0; int OUT_WIDTH = 0; IdTableView<WIDTH> = detail::IdTableTemplated<0, detail::IdTableViewWrapper<ad_utility::AllocatorWithLimit<ValueId> >, ad_utility::AllocatorWithLimit<ValueId> >; size_t = long unsigned int; IdTableView<R_WIDTH> = detail::IdTableTemplated<0, detail::IdTableViewWrapper<ad_utility::AllocatorWithLimit<ValueId> >, ad_utility::AllocatorWithLimit<ValueId> >; IdTableStatic<OUT_WIDTH> = detail::IdTableTemplated<0, detail::IdTableVectorWrapper<ad_utility::AllocatorWithLimit<ValueId> >, ad_utility::AllocatorWithLimit<ValueId> >]’:
qlever-code/src/engine/Join.cpp:541:22:   required from ‘void Join::join(const IdTable&, size_t, const IdTable&, size_t, IdTable*) [with int L_WIDTH = 0; int R_WIDTH =0; int OUT_WIDTH = 0; size_t = long unsigned int]’
qlever-code/src/engine/Join.cpp:149:3:   required from here
qlever-code/src/engine/Join.cpp:710:61: error: cannot call member function ‘void Join::addCombinedRowToIdTable(std::decay_t<typename detail::IdTableTemplated<OTHER_COLS, detail::IdTableVectorWrapper<ad_utility::AllocatorWithLimit<ValueId> >, ad_utility::AllocatorWithLimit<ValueId> >::const_row_type>&, std::decay_t<typename detail::IdTableTemplated<WIDTH, detail::IdTableVectorWrapper<ad_utility::AllocatorWithLimit<ValueId> >, ad_utility::AllocatorWithLimit<ValueId> >::const_row_type>&, size_t, IdTableStatic<TABLE_WIDTH>*) const [with int A_WIDTH = 0; int B_WIDTH = 0; int TABLE_WIDTH = 0; std::decay_t<typename detail::IdTableTemplated<OTHER_COLS, detail::IdTableVectorWrapper<ad_utility::AllocatorWithLimit<ValueId> >, ad_utility::AllocatorWithLimit<ValueId> >::const_row_type> = std::decay<detail::ConstRow>::type; typename detail::IdTableTemplated<OTHER_COLS, detail::IdTableVectorWrapper<ad_utility::AllocatorWithLimit<ValueId> >, ad_utility::AllocatorWithLimit<ValueId>>::const_row_type = detail::ConstRow; std::decay_t<typename detail::IdTableTemplated<WIDTH, detail::IdTableVectorWrapper<ad_utility::AllocatorWithLimit<ValueId> >, ad_utility::AllocatorWithLimit<ValueId> >::const_row_type> = std::decay<detail::ConstRow>::type; typename detail::IdTableTemplated<WIDTH, detail::IdTableVectorWrapper<ad_utility::AllocatorWithLimit<ValueId> >, ad_utility::AllocatorWithLimit<ValueId> >::const_row_type = detail::ConstRow; size_t = long unsigned int; IdTableStatic<TABLE_WIDTH> = detail::IdTableTemplated<0, detail::IdTableVectorWrapper<ad_utility::AllocatorWithLimit<ValueId> >,ad_utility::AllocatorWithLimit<ValueId> >]’ without object
```

I can't really see, **why** it shouldn't work, but I would suspect `CALL_FIXED_SIZE3` because of the `without object`.

Reply:
The problem here is that
`Join::join` is a non-static member function, and `addCombinedRow` is also a non-static member function.

But `doGallopInnerJoin` is a static member function, so it cannot call a non-static function.

Maybe the easiest way to make this work is to make the `doGallop.....` functions also non-static (in the .h-file remove the "static" and simply add `const` afterwards. Or, probably better, make the `add...` function static (if that is possible).

Reply:
There is the timeout mechanism (`checkTimeout` `checkTimeoutAfterNCalls` or whatever these functions are called.). Those are not static as far as I remember. So the functions being not static is not a big problem.

And yes: The static keyword only being allowed and present in the header is often confusing, I agree.

Reply:
Once again, I should have read the code more carefully. Thank you for pointing that out!

Reply:
Added more use of the helper function in [8cc8dc2](https://github.com/ad-freiburg/qlever/pull/821/commits/8cc8dc2388cefb7bde6b4ecfbd23d981015b7453).